apollo-federation-types/src/build_plugin/plugin_result.rs

```diff
@@ -0,0 +1,157 @@
+use super::BuildMessage;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct PluginResult {
+    pub is_success: bool,
+    pub schema: Option<String>,
+    pub build_messages: Vec<BuildMessage>,
+
+    /// Other untyped JSON included in the build output.
+    #[serde(flatten)]
+    other: crate::UncaughtJson,
+}
+
+impl PluginResult {
+    /**
+    We may succed in Rust's perspective, but inside the JSON message may be isSuccess: false
+    and buildMessages from composition telling us what went wrong.
+
+    If there are, promote those to more semantic places in the output object.
+    If there are not, cooool, pass the data along.
+    */
+    pub fn from_plugin_result(result_json: &str) -> Self {
+        let serde_json: Result<PluginResult, serde_json::Error> = serde_json::from_str(result_json);
+        match serde_json {
+            Ok(js_response) => js_response,
+            Err(json_error) => Self {
+                is_success: false,
+                schema: None,
+                build_messages: BuildMessage::to_build_errors(
+                    vec![format!(
+                        "Could not parse JSON from Rust. Received error {}",
+                        json_error
+                    )],
+                    Some("PLUGIN_EXECUTION".to_string()),
+                    Some("PLUGIN_EXECUTION".to_string()),
```

Comment:
what's going on here?

Reply:
We use `code` and `step` to highlight display information in studio. I'm trying to indicate that this is a plugin execution error as opposed to a error from the plugin itself.

I was thinking in studio we can build alerting on these kinds of errors as well as give better tool tip information if these appear. 

Ideally this could be an enum, but I don't think we have an exhaustive list of codes or steps anywhere, and we haven't really discussed how we want people to use them programatically.

Reply:
Makes sense - kinda confusing to see both the code and the step be the same exact string.

I think your instinct to make an enum is OK, and honestly I think it's fine that we don't know 100% what it will always be. You can tag enums with `#[non_exhaustive]` in Rust and then all consuming code has to handle the `_` case (something was passed that is not represented by a variant in this enum). I think it would make sense to have two separate enums `Code` and `Step` so you can do `Code::PluginExecution.to_string()` and `Step::PluginExecution.to_string()` and get some sweet readability improvements. You can get the `.to_string()` to work with an `impl Display for Code` and `impl Display for `Step` blocks, VS code should auto-complete the function block for you, but here's an example: https://doc.rust-lang.org/rust-by-example/hello/print/print_display.html